Quasar Vuejs Google authentication with firebase (#2459)
section-engineering-education/engineering-education: articles/quasar-google-auth-with-firebase/index.md

4. After the project is ready click `Continue`. It will take you to the project overview. Click the `web` icon, then register an app name, name it `quasar-firebase-googleauth` and click `Register App`
5. You'll be taken to `step 2 - Add Firebase SDK`, copy the whole script.

We're going to install firebase and create a boot file to initialize firebase before our app runs.
Go back to our project, open the terminal, run the command below:

```bash
yarn add firebase
quasar new boot firebase
```

Navigate to `quasar.conf.js` file, search for `boot` it should be an empty array. Add `firebase.js` boot file in it as indicated below.

```JavaScript
boot: ['firebase'],
```

Navigate to the `boot` folder, open `firebase.js` file. 

Paste the script you copied the previous step for the firebase SDK. `firebase.js` file structure should resemble the one below.

```JavaScript
import firebase from "firebase";

const firebaseConfig = {
    apiKey: "",
    authDomain: "",
    projectId: "",
    storageBucket: "",
    messagingSenderId: "",
    appId: ""
  };
  firebase.initializeApp(firebaseConfig);

export default firebase
```

### Access Control for Routes

> We're going to disable access to routes which requires authorization when a user needs to access them. This is quite easy. Navigate to `routes` folder, open `index.js` file.

Reviewer: ```suggestion
We're going to disable access to routes that require authorization when a user needs to access them. This is quite easy. Navigate to the `routes` folder, open the `index.js` file.
```

> `Import firebase from "firebase"` at the top of the file just like we did in the `firebase.js` boot file. When done, add the following code just above the line `return Router`:

Reviewer: ```suggestion
Add `import firebase from "firebase"` at the top of the file just like we did in the `firebase.js` boot file. When done, add the following code just above the line `return Router`:
```

```JavaScript
  Router.beforeEach(async (to, from, next) => {
    const auth = to.meta.requiresAuth
    if (auth && !await firebase.getCurrentUser()) {
      next('/');
    } else {
      next();
    }
  })
```

> Now if the currentUser is null or undefined, we should redirect users to the auth path (`/`). But how do we get currentUser? We can’t use `firebase.auth().currentUser` because on page refresh that property has not been set yet before the requiresAuth guard is triggered. 

Reviewer: ```suggestion
Now if the currentUser is null or undefined, we should redirect users to the auth path (`/`). But how do we get currentUser? We can’t use `firebase.auth().currentUser` because on page refresh that property has not been set yet before the `requiresAuth` guard is triggered. 
```

> We will have to use the onAuthStateChanged callback somehow. We have to add a method to the firebase object after we initialize the firebase app. The method is added to `firebase.js` boot file. Edit the firebase boot file to resemble the code below.

Reviewer: ```suggestion
We'll have to use the `onAuthStateChanged` callback somehow. We have to add a method to the firebase object after we initialize the firebase app. The method is added to the `firebase.js` boot file. Edit the firebase boot file to resemble the code below.
```

```JavaScript
const firebaseConfig = {
    apiKey: "",
    authDomain: "",
    projectId: "",
    storageBucket: "",
    messagingSenderId: "",
    appId: ""
  };
  firebase.initializeApp(firebaseConfig);

  firebase.getCurrentUser = () => {
    return new Promise((resolve, reject) => {
      const unsubscribe = firebase.auth().onAuthStateChanged(user => {
        unsubscribe();
        resolve(user);
      }, reject);
    })
  };
```  

> The method after `firebase.initializeApp(firebaseConfig);` will return a Promise which resolves currentUser as soon as it is set. `onAuthStateChanged` will trigger the callback immediately with either null or the user object if signed in. Then we unsubscribe to not listen for further changes.

Reviewer: ```suggestion
`firebase.getCurrentUser` will return a Promise which resolves currentUser as soon as it is set. `onAuthStateChanged` will trigger the callback immediately with either null or the user object if signed in. Then we unsubscribe to not listen for further changes.
```

### Activate SignIn Methods

We need to activate signin methods provided by firebase, we'll activate the `google` and `Email/Password` providers. Go back to the firebase console where we copied the firebase SDK script. click `Continue to console`, it will take you to the project overview.

On the left sidebar click `Authentication`, then click `Set up sign-in method`. Click on `Email/Password` and `Google` providers to activate them. For Email/Password click the first toggle button, don't enable the `passwordless sigin-in` option, while for the Google provider make sure to fill in your `Project support email ` preferably your email address.

> Once done, we can get started adding functionalities on our view.

Reviewer: Can you provide helpful screenshots for these actions? It'll be easier for the readers to follow along. You can highlight the buttons that the user should click to make it even more easier for the reader to follow.

Open `AuthComponent.vue` file, in the `createUser()` method, this method will be used to create a new user to the database, paste the following code for the method.

```JavaScript
createUser(email, password) {
  firebase.auth().createUserWithEmailAndPassword(email, password)
    .then(auth => {
      this.$q.notify({message: 'Sign In Success.'})
      this.$router.push('/home')
    })
    .catch(error => {console.log(error)
    })
},
```

When a user is created it redirects the user to the home page, with `Sign In Success` notification at the bottom of the page.

Next, we're going to create a `signInExistingUser()` method, this signs in already registered users. Paste the following code for the method.

```JavaScript
signInExistingUser (email, password) {
  firebase.auth().signInWithEmailAndPassword(email, password)
    .then((userCredential) => {
      this.$q.notify({message: 'Sign In Success.'})
      this.$router.push('/home')
    })
    .catch(error => { console.log(error)})
},
```

By now, you can sign-in with google or with email/password. If successful, you are redirected to the home page, containing a welcome message and a logout button.

> If you're working locally, and have encountered a problem signing in, go to firebase console in the project, click `Authentication`, then `Sign-in method`, below `Sign-in providers`, you'll see `Authorized domains`. Click `Add domain` and add `localhost`. You can add custom domains when your app is hosted.

Reviewer: Screenshots would be helpful.
